# Decision Tree Linear Regression
# .




Afin de créer un arbre de décision, nous allons étudier un CSV sur les revenus d'individus aux USA.
Cela contient les informations sur le status marital, l'âge, le type d'emploi etc.
Les datas sont de 1994. Nous voulons predire si le salaire sera inferieur voire égal à 50k ou supérieur à 50k

Voilà la source : [University of California, Irvine's website](http://archive.ics.uci.edu/ml/datasets/Adult)
Le CSV est inclus.

### <span style="color:purple">Importe le fichier income.csv fourni
 </span>

In [5]:
import pandas

income = pandas.read_csv("income.csv", index_col=False)
income.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country high_income  
0          2174             0              40   United-States       <=50K  
1             0             0              13   United-States       <=50K  
2             0             0              40   United-States       <=50K  
3             0             0              40   United-States       <=50K  
4             0             0              40            Cuba       <=50K

 ### <span style="color:purple">Combien de variables y a t-il ?
 </span>

 <span style="color:green">Attendu : 32561</span>

## Catégories

### <span style="color:purple">Prends le temps d'interpreter les datas. Quels sont les differents facteurs, les	Θ. Sous quelle forme sont elles classifiées. Sex donne "Male" ou "Female". "workclass" contient plusieurs sortes de réponses.   Afin de pouvoir les analyser nous allons les convertirs en valeures numériques en fonction des categories.   Pour cela utilise la class [pandas.Categorical()](http://pandas.pydata.org/pandas-édocs/stable/generated/pandas.Categorical.html#pandas.Categorical)</span>

In [6]:
#income.head()

### <span style="color:purple">Convertis le reste des catgorical columns dans `income` (education, marital_status, occupation, relationship, race, sex, native_country, and high_income) en categories numerique.</span>


In [8]:
to_convert = ["education","marital_status","occupation","relationship","race","sex","native_country","high_income"]

income.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country high_income  
0          2174             0              40   United-States       <=50K  
1             0             0              13   United-States       <=50K  
2             0             0              40   United-States       <=50K  
3             0             0              40   United-States       <=50K  
4             0             0              40            Cuba       <=50K

### <span style="color:purple">A present divise income en deux `DataFrames` afin de diviser ceux travaillant dans le secteur privé ou non. </span>
*python hint : Booleans to split a dataframe.*

### <span style="color:purple">Quelle est la proportion de `private_income` et de `public_incomes` ?
 </span>

In [10]:

#print("private_incomes_proportion",private_incomes_prop)
#print("public_incomes_proportion",public_incomes_prop)

 <span style="color:green">private_incomes 0.6970301894904948 <br>
public_incomes 0.3029698105095052</span>

## Entropy

###  <span style="color:purple"> Calcule l'entropie de `high_income` avec les separations donnees. </span>

$$-\sum _{ i=1 }^{ c }{ P({ x }_{ i })\log _{ b }{ P({ x }_{ i }) }  } $$

In [11]:
import math

#print(entropy)

 <span style="color:green">Attendu : 0.8848909340231308</span>

 ### <span style="color:purple">Quelle est la proportion de `high_income` et de `low_incomes` ?
 </span>

In [12]:
#print("prob_0 :", prob_0)
#print("prob_1 :", prob_1)

 <span style="color:green">Attendu :<br>prob_0 : 0.7591904425539756 <br>
prob_1 : 0.2408095574460244</span>

 ### <span style="color:purple">Quelle est l'entropie `high_income` et de `low_incomes` ?
 </span>

In [13]:
#print(income_entropy)

 <span style="color:green">Attendu : 0.7963839552022132</span>

### <span style="color:purple">Code une fonction généraliste pour l'Entropie et vérfie que tu obtiens le même résultat. Puis vérifie que la réponse coïncide</span>
 
 *hint : numpy.bincount*

In [14]:
import numpy

def calc_entropy(column):

    # Détermine le nombre d'occurence de chacune des valeurs dans la colonne

    # détermine la probabilité d'occurence de chacune des catégories

    
    entropy = 0
    # fais une boucle par probabilités et additionne les résultats


print('high_income', calc_entropy(income['high_income']))
print('workclass', calc_entropy(income['workclass']))

high_income None
workclass None


 <span style="color:green">Attendu : <br> high_income 0.796383955202 <br>
workclass 1.64797692751</span>

###  <span style="color:purple">Voila un cas avec plus de 2 possibilites, crée deux subset de `age` en fonction du médian (l'un incluant et l'autre excluant)</span>

In [15]:
#left_split.head(3)

In [17]:
#right_split.head(3)

## Information Gain

###  <span style="color:purple">trouve le médian de `age`</span>
 

In [20]:
median_age

37.0

 <span style="color:green">Attendu: 37.0</span>

###  <span style="color:purple">Calcule le `Information Gain` de `age` a partir de la separation high_income</span>
 
$$Entropy(T)\quad =\sum _{ i=1 }^{ c }{ P({ x }_{ i })\log _{ b }{ P({ x }_{ i }) }  } \\ \\ IG(T,\quad A)\quad =\quad Entropy(T)-\sum _{ v\epsilon A }^{  }{ \frac { |{ T }_{ v }| }{ |T| }  } .Entropy({ T }_{ v })$$


In [24]:
income_entropy = calc_entropy(income["high_income"])
#calculez l' `age_information_gain`


 <span style="color:green">Attendu: -0.00361604479779</span>

 ### <span style="color:purple">Créer une fonction `calc_information_gain`</span>

In [92]:
def calc_information_gain(data, split_name, target_name):

    # calcule l'entropy d'origine
    original_entropy = calc_entropy(data[target_name])
    
    # trouve le médiant de la colonne
    column = 
    median = 
    
    # crée deux subset en fonction du médian
    
    # calcule le subset entropy de chacun des set avec une boucle
    to_subtract = 0
    
    
    # Retourne information gain
    return original_entropy - to_subtract

# Vérifie que la valeur coïncide avec `income`, "age", "high_income"
print(calc_information_gain(income, "age", "high_income"))

0.0470286613047


 <span style="color:green">Attendu: 0.0470286613047</span>

 ### <span style="color:purple">Puis une `liste information_gains` de toutes les colonnes</span>

In [93]:
columns = ["age", "workclass", "education_num", "marital_status", "occupation", "relationship", "race", "sex", "hours_per_week", "native_country"]
information_gains = []
#faire une boucle
for col in columns:
    information_gain = calc_information_gain(income, col, "high_income")
    information_gains.append(information_gain)

 ### <span style="color:purple">Selectionne le nom de colonne avec la valeur la plus élevée</span>

In [94]:
highest_gain_index = information_gains.index(max(information_gains))
highest_gain = columns[highest_gain_index]
highest_gain

'marital_status'